### PR TITLE
Lookup join

### PR DESCRIPTION
**Pull request by cube2222**

### CODE REVIEW

File: storage/json/datasource.go [hunk 1]
```diff
@@ -140,5 +140,5 @@ func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
 		return fields[i] < fields[j]
 	})
 
-	return execution.NewRecord(fields, aliasedRecord), nil
+	return execution.NewRecord(fields, aliasedRecord, execution.WithID(execution.NewID(execution.GetRawStreamID().Id))), nil
```
Comment (cube2222) [inline storage/json/datasource.go]: Need to change this in all datasources.

File: execution/lookup_join.go [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+				return nil, errors.Wrapf(err, "couldn't delete job with recordID %s from jobs list", jobRecordID.Show())
+			}
+
+			if err := recordsQueue.Clear(); err != nil {
+				return nil, errors.Wrapf(err, "couldn't clear joined record queue for job with recordID %s", jobRecordID.Show())
+			}
+
+			err := rs.HandleControlMessages(ctx, tx)
+			if err == ErrEndOfStream {
+				if err := iter.Close(); err != nil {
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/lookup_join.go]: you wouldn't handle the error then, and using return value changing in deferred functions is really shady. (it's possible)

File: execution/output_queue.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+var queueElementsPrefix = []byte("$queue_elements$")
+
+func (q *OutputQueue) Push(ctx context.Context, element proto.Message) error {
+	queueElements := storage.NewDeque(q.tx.WithPrefix(queueElementsPrefix))
+
+	err := queueElements.PushBack(element)
+	if err != nil {
+		return errors.Wrap(err, "couldn't append element to queue")
+	}
+	return nil
```
Comment (woojtek) [inline execution/output_queue.go]: newline here? I don't know if its necessary but I always write newline before last return :P

File: execution/output_queue.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+	return nil
+}
+
+func (q *OutputQueue) Pop(ctx context.Context, msg proto.Message) error {
+	queueElements := storage.NewDeque(q.tx.WithPrefix(queueElementsPrefix))
+
+	err := queueElements.PopFront(msg)
+	if err == storage.ErrNotFound {
+		subscription := q.tx.GetUnderlyingStorage().Subscribe(ctx)
```
Comment (woojtek) [inline execution/output_queue.go]: Why is that needed? Could You write 2 sentences about it?

File: execution/process.go [hunk 5]
```diff
@@ -197,59 +198,19 @@ func (p *ProcessByKey) MarkEndOfStream(ctx context.Context, tx storage.StateTran
...
-
-var queueElementsPrefix = []byte("$queue_elements$")
-
-func (q *OutputQueue) Push(ctx context.Context, element proto.Message) error {
-	queueElements := storage.NewDeque(q.tx.WithPrefix(queueElementsPrefix))
-
-	err := queueElements.PushBack(element)
+func (p *ProcessByKey) MarkError(ctx context.Context, tx storage.StateTransaction, err error) error {
+	outputQueue := NewOutputQueue(tx.WithPrefix(outputQueuePrefix))
+	err = outputQueue.Push(ctx, &QueueElement{Type: &QueueElement_Error{Error: err.Error()}})
```
Comment (woojtek) [inline execution/process.go]: Can you indent QueueElement here like You did in kafka for example? It looks better imo

File: execution/record.go [hunk 6]
```diff
@@ -228,6 +234,34 @@ func (r *Record) Hash() (uint64, error) {
 	return hashstructure.Hash(append(values, fields...), nil)
 }
 
+// GetRandomRecordID can be used to get a new random RecordID.
+func NewRecordID(id string) *RecordID {
```
Comment (woojtek) [inline execution/record.go]: Is this exactly the same as `NewId` function from the beginning of this file? :P

File: execution/tvf/range.go [hunk 7]
```diff
@@ -48,28 +51,45 @@ func (r *Range) Get(ctx context.Context, variables octosql.Variables, streamID *
 		return nil, nil, errors.Errorf("invalid range start point: %v", end)
 	}
 
+	currentIndexState := storage.NewValueState(storage.GetStateTransactionFromContext(ctx).WithPrefix(streamID.AsPrefix()).WithPrefix(currentIndexPrefix))
```
Comment (woojtek) [inline execution/tvf/range.go]: Afaik we've always declared `tx := storage.GetStateTransactionFromContext(ctx)` separately, but I guess that's OK since we use that only for this value state

File: execution/tvf/range.go [hunk 8]
```diff
@@ -48,28 +51,45 @@ func (r *Range) Get(ctx context.Context, variables octosql.Variables, streamID *
...
 
 type RangeStream struct {
-	current, endExclusive int
+	streamID     *execution.StreamID
+	endExclusive int
 }
 
 func (s *RangeStream) Next(ctx context.Context) (*execution.Record, error) {
-	if s.current >= s.endExclusive {
+	currentIndexState := storage.NewValueState(storage.GetStateTransactionFromContext(ctx).WithPrefix(s.streamID.AsPrefix()).WithPrefix(currentIndexPrefix))
```
Comment (woojtek) [inline execution/tvf/range.go]: Same

File: storage/json/datasource.go [hunk 9]
```diff
@@ -81,6 +82,8 @@ type RecordStream struct {
 	decoder                       *json.Decoder
 	isDone                        bool
 	alias                         string
+	offset                        int
+	streamID                      *execution.StreamID
```
Comment (woojtek) [inline storage/json/datasource.go]: streamID as a first field like excel/csv ?

File: execution/lookup_join.go [hunk 10]
```diff
@@ -0,0 +1,726 @@
...
+
+	return nil
+}
+
+func (rs *LookupJoinStream) AddRecord(ctx context.Context, tx storage.StateTransaction, inputIndex int, record *Record) error {
+	var toBeJoinedQueue = NewOutputQueue(tx.WithPrefix(toBeJoinedQueuePrefix))
+
+	err := toBeJoinedQueue.Push(ctx, &QueueElement{
+		Type: &QueueElement_Record{
+			Record: record,
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/lookup_join.go]: This is a lookup join, not a stream join. We only get records from one stream into this queue. Then there's a separate queue for each lookup join job.

File: execution/lookup_join.go [hunk 11]
```diff
@@ -0,0 +1,726 @@
...
+}
+
+func (rs *LookupJoinStream) GetWatermark(ctx context.Context, tx storage.StateTransaction) (time.Time, error) {
+	outputWatermarkState := storage.NewValueState(tx.WithPrefix(outputWatermarkPrefix))
+	var octoWatermark octosql.Value
+	err := outputWatermarkState.Get(&octoWatermark)
+	if err == storage.ErrNotFound {
+		return time.Time{}, nil
+	} else if err != nil {
+		return time.Time{}, errors.Wrap(err, "couldn't get output watermark")
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/lookup_join.go]: I don't understand

File: execution/lookup_join.go [hunk 12]
```diff
@@ -0,0 +1,726 @@
...
+	if err != nil {
+		return errors.Wrap(err, "couldn't push record to queue of records to be joined")
+	}
+
+	return nil
+}
+
+func (rs *LookupJoinStream) Next(ctx context.Context, tx storage.StateTransaction) (*Record, error) {
+	// Check if this stream isn't done already.
+	var endOfStreamState = storage.NewValueState(tx.WithPrefix(endOfStreamPrefix))
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/lookup_join.go]: no, somewhere else in this package, it's used often.

File: execution/lookup_join.go [hunk 13]
```diff
@@ -0,0 +1,726 @@
...
+	var phantom octosql.Value
+	var err error
+	// Go over all jobs in order and try to get records to return from any of them.
+	for err = iter.Next(&jobRecordID, &phantom); err == nil; err = iter.Next(&jobRecordID, &phantom) {
+		sourceRecordState := storage.NewValueState(tx.WithPrefix(jobRecordID.AsPrefix()).WithPrefix(sourceRecordPrefix))
+		recordsQueue := storage.NewDeque(tx.WithPrefix(outputPrefix).WithPrefix(jobRecordID.AsPrefix()))
+
+		var outputElement QueueElement
+
+		if rs.isLeftJoin {
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/lookup_join.go]: described in real life.

File: execution/lookup_join.go [hunk 14]
```diff
@@ -0,0 +1,728 @@
...
+
+		// Save that we've done the initialization routine.
+		phantom = octosql.MakePhantom()
+		err := isInitialized.Set(&phantom)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "couldn't save initialization status of lookup join")
+		}
+	} else if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get initialization status of lookup join")
+	} else if err == nil {
```
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/lookup_join.go]: If it's not initialized then there are no jobs, so you don't have anything to run workers for.

File: execution/lookup_join.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (woojtek) [reply execution/lookup_join.go]: nil instead of time.Time{} when returning error I mean

File: execution/lookup_join.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (JasiekChomiak) [reply execution/lookup_join.go]: Oh because the initialization happens in the `jobTokenQueuePrefix` storage, and jobs are taken from `jobsPrefix` storage - didn't notice, sorry.

File: execution/lookup_join.go [hunk 15]
```diff
@@ -613,7 +615,7 @@ func (rs *LookupJoinStream) GetWatermark(ctx context.Context, tx storage.StateTr
 	if err == storage.ErrNotFound {
 		return time.Time{}, nil
 	} else if err != nil {
-		return time.Time{}, errors.Wrap(err, "couldn't get output watermark")
+		return time.Time{}, errors.Wrap(err, "couldni trzeba't get output watermark")
```
Comment (JasiekChomiak) [inline execution/lookup_join.go]: Something went wrong here ;)

Comment (woojtek) [inline execution/lookup_join.go]: Refactor bug :)

File: execution/sourcestorage.go [hunk 16]
```diff
@@ -23,8 +23,13 @@ var inputStreamIDPrefix = []byte("$input$")
 // GetRawStreamID can be used to get a new StreamID without saving it.
 func GetRawStreamID() *StreamID {
 	id := ulid.MustNew(ulid.Timestamp(time.Now()), rand.Reader)
+	return NewStreamID(id.String())
+}
+
+// GetRawStreamID can be used to get a new StreamID without saving it.
```
Comment (JasiekChomiak) [inline execution/sourcestorage.go]: Is it on purpose that this comment appears above `GetRawStreamID` and `NewStreamID` and says exactly the same thing?

File: execution/lookup_join.go (hunk 15 above)
Comment (cube2222) [reply execution/lookup_join.go]: :o

File: execution/sourcestorage.go (hunk 16 above)
Comment (cube2222) [reply execution/sourcestorage.go]: no, copy-paste, thank

File: execution/lookup_join.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (cube2222) [reply execution/lookup_join.go]: no, because this returns time.Time, not *time.Time
